server/aesdsocket.c

```diff
@@ -0,0 +1,242 @@
+	of_memsz = (of_sz/PAGE_SIZE + 1)*PAGE_SIZE;
+	s = ftruncate(ofd, of_memsz);
+	if (s == -1) cleanup(SRC_FTRUNCATE);
+	void* of_mem = mmap(NULL, of_memsz, PROT_READ | PROT_WRITE, MAP_SHARED, ofd, 0);
+	if (of_mem == MAP_FAILED) cleanup(SRC_MMAP);
+	char* of_pt = (char*) of_mem + of_sz;
+
+	// create socket
+	asock = socket(AF_INET, SOCK_STREAM, 0);
+	if (asock == -1) cleanup(SRC_SOCKET);
+	int opt = 1;
+	s = setsockopt(asock, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(int)); // skip waiting for address to be freed
+	if (s == -1) cleanup(SRC_SOCKET);
+
+	// bind socket
+	struct sockaddr_in serv_addr = {
+		.sin_family = AF_INET,
+		.sin_port = htons(9000),
+	//	.sin_addr = htonl(INADDR_LOOPBACK)
+		.sin_addr = htonl(INADDR_ANY)
+	};
+	s = bind(asock, &serv_addr, sizeof(struct sockaddr_in));
+	if (s == -1) cleanup(SRC_BIND);
+
+	// alloc packet buffer
+	size_t max_packet = INITIAL_MAX_PACKET;
+	packet = malloc(max_packet);
+	if (!packet) cleanup(SRC_MALLOC);
+
+	// daemonise if requested
+	if (daemon) {
+		pid_t cpid = fork();
+		if (cpid == 0) { // in daemon
+			int nullfd = open("/dev/null", O_RDWR);
+			if (nullfd == -1) cleanup(SRC_DUP);
+			int fd = dup2(nullfd, fileno(stdout));
+			if (fd == -1) cleanup(SRC_DUP);
+			fd = dup2(nullfd, fileno(stderr));
+			if (fd == -1) cleanup(SRC_DUP);
+			fd = dup2(nullfd, fileno(stdin));
+			if (fd == -1) cleanup(SRC_DUP);
+			
+			syslog(LOG_INFO, "daemonised with pid %d, sid %d", getpid(), setsid()); 
+		} else {
+			printf("Daemon pid %d\n", cpid);
+		   	exit(0);
+		}
+	}	
+
+	// listen on socket
+	s = listen(asock, MAX_BACKLOG);
+	if (s == -1) cleanup(SRC_LISTEN);
+
+	// main server loop (exited via interrupt)
+	while (1) {
+		// wait for connection
+		struct sockaddr_in cli_addr = {0};
+		socklen_t addrlen = sizeof(struct sockaddr_in);
+		csock = accept4(asock, &cli_addr, &addrlen, 0);
+		if (csock == -1) cleanup(SRC_ACCEPT);
+		assert(addrlen <= sizeof(struct sockaddr_in));
+
+		// log accepted connection
+		const char* cip = inet_ntoa(cli_addr.sin_addr);
+		syslog(LOG_INFO, "Accepted connection from %s\n", cip);
+
+		// get packet
+		char* delim = NULL;
+		size_t packet_sz = 0;
+		size_t max_read = max_packet;
+		while (!delim) {
+			assert(packet_sz <= max_packet);
+			ssize_t read_sz = read(csock, &packet[packet_sz], max_read);	
+			if (read_sz == -1)
+				cleanup(SRC_READ);
+			
+			if (read_sz == max_read) { // need to expand buffer
+				max_read = max_packet;
+				max_packet *= 2;
+				packet = realloc(packet, max_packet);
+				if (!packet) cleanup(SRC_REALLOC);
+			}
+
+			delim = memchr(&packet[packet_sz], '\n', read_sz);
+			packet_sz += read_sz;
+			max_read -= read_sz;
+		}
+
+		// log packet
+		if (of_sz + packet_sz > of_memsz) { // need to expand mapping
+			void* prev_mem = of_mem;
+			size_t new_memsz = ((of_sz + packet_sz)/PAGE_SIZE + 1)*PAGE_SIZE;
+			s = ftruncate(ofd, new_memsz);
+			if (s == -1) cleanup(SRC_FTRUNCATE);
+			of_mem = mremap(of_mem, of_memsz, new_memsz, MREMAP_MAYMOVE);
```

Comment:
@georgehodgkins what assumption does this make about the total size of RAM required relative to the total number of bytes in all packets sent?